wandb/sdk/launch/builder/build.py

```diff
@@ -44,6 +45,68 @@
 _DEFAULT_DOCKERFILE_NAME = "Dockerfile.wandb"
 
 
+def registry_from_uri(uri: str) -> AbstractRegistry:
+    """Create a registry helper object from a uri.
+
+    This function parses the URI and determines which supported registry it
+    belongs to. It then creates a registry helper object for that registry.
+    The supported remote registry types are:
+    - Azure Container Registry
+    - Google Container Registry
+    - AWS Elastic Container Registry
+
+    The format of the URI is as follows:
+    - Azure Container Registry: <registry-name>.azurecr.io/<repo-name>/<image-name>
+    - Google Container Registry: <location>-docker.pkg.dev/<project-id>/<repo-name>/<image-name>
+    - AWS Elastic Container Registry: <registry-name>.dkr.ecr.<region>.amazonaws.com/<repo-name>/<image-name>
+
+    Our classification of the registry is based on the domain name. For example,
+    if the uri contains `.azurecr.io`, we classify it as an Azure
+    Container Registry. If the uri contains `.dkr.ecr`, we classify
+    it as an AWS Elastic Container Registry. If the uri contains
+    `-docker.pkg.dev`, we classify it as a Google Artifact Registry.
+
+    This function will attempt to load the approriate cloud helpers for the
+
+    `https://` prefix is optional for all of the above.
+
+    Arguments:
+        uri: The uri to create a registry from.
+
+    Returns:
+        The registry.
+
+    Raises:
+        LaunchError: If the registry helper cannot be loaded for the given URI.
+    """
+    if uri.startswith("https://"):
+        uri = uri[len("https://") :]
+
+    if "azurecr.io" in uri:
+        from wandb.sdk.launch.registry.azure_container_registry import (
+            AzureContainerRegistry,
+        )
+
+        return AzureContainerRegistry(uri=uri)
+
+    elif "docker.pkg.dev" in uri:
```

Comment:
If the environment still exists, why not use that?

Reply:
I can improve this to match the location of the domain in the pattern instead of just looking it in the string. The main goal of this PR is to prevent users from having to enter this registry block with a code for the registry type, and this is only possible because we can infer the helper to load based on the registry uri format. So if that inference is deemed unsafe this PR is sort of out the window. I don't want to infer based on environment either because I also aim to make the environment block optional and redundant. The only thing that remains in the environment block is a region and we can get the region if needed from an environment variable or in cluster metadata server in like every case.